executor/src/witgen/machines/block_machine.rs

```diff
@@ -319,174 +321,90 @@ impl<'a, T: FieldElement> BlockMachine<'a, T> {
             ))
         }
     }
 
-    fn compute_updates(
-        &self,
-        row: DegreeType,
-        fixed_data: &'a FixedData<T>,
-        left: &[AffineExpression<&'a PolynomialReference, T>],
-        right: &'a SelectedExpressions<T>,
-        identity: IdentityInSequence,
-        identity_processor: &mut IdentityProcessor<'a, '_, T>,
-    ) -> EvalResult<'a, T> {
-        match identity {
-            IdentityInSequence::Internal(i) => identity_processor.process_identity(
-                self.identities[i],
-                &self.get_current_row_pair(row, fixed_data),
-            ),
-            IdentityInSequence::OuterQuery => {
-                assert!(row as usize % self.block_size == self.block_size - 1);
-                identity_processor.process_link(
-                    left,
-                    right,
-                    &self.get_current_row_pair(row, fixed_data),
-                )
-            }
-        }
-    }
-
-    fn apply_updates(
-        &mut self,
-        row: DegreeType,
-        identity: IdentityInSequence,
-        updates: &EvalValue<&'a PolynomialReference, T>,
-        left_mut: &mut [AffineExpression<&'a PolynomialReference, T>],
-    ) -> bool {
-        if updates.constraints.is_empty() {
-            return false;
+    /// Takes a block of rows, which contains the last row of its previous block
+    /// and the first row of its next block. The first row of its next block is ignored,
+    /// the last row of its previous block is merged with the one we have already.
+    /// This is necessary to handle non-rectangular block machines, which already use
+    /// unused cells in the previous block.
+    fn append_block(&mut self, mut new_block: Vec<Row<'a, T>>) -> Result<(), EvalError<T>> {
+        if self.rows() + self.block_size as DegreeType >= self.fixed_data.degree {
+            return Err(EvalError::RowsExhausted);
         }
 
-        match identity {
-            IdentityInSequence::Internal(index) => {
-                log::trace!("    Updates from: {}", self.identities[index])
-            }
-            IdentityInSequence::OuterQuery => log::trace!("    Updates from: outer query"),
-        };
-
-        let (before, after) = self.data.split_at_mut(row as usize + 1);
-        let current = before.last_mut().unwrap();
-        let next = after.first_mut().unwrap();
-
-        let mut row_updater = RowUpdater::new(current, next, row);
-        for (poly, c) in &updates.constraints {
-            if self.witness_cols.contains(&poly.poly_id()) {
-                row_updater.apply_update(poly, c);
-            } else if let Constraint::Assignment(v) = c {
-                for l in left_mut.iter_mut() {
-                    log::trace!("      => {} (outer) = {}", poly, v);
-                    l.assign(poly, *v);
+        assert_eq!(new_block.len(), self.block_size + 2);
+
+        // 1. Ignore the first row of the next block:
```

Comment:
Shouldn't we at least store the values that are not unknown?

Reply:
I think it's not necessary unless we want to support blocks accessing cells after the latch row (similar to how we allow them to access cells before the first row). But supporting more block shapes is kind of a different topic (e.g., then we should also generalize it support having the latch on any row, not just the last).

If the values that are removed here are already for the next block, they will just be computed again when the next block is computed. The machine will only have more context, no information is lost.

Reply:
Ah ok, I thought these are values that can only be computed from the previous row, but you are right, they might be computed from the same row.

Then let's leave it like that for now!